### ISSUE TITLE
Make spring-hateoas-webmvc module compatible with spring-hateoas 1.3.4

### ISSUE THREAD

**Issue by ask4gilles** [Blazebit/blaze-persistence#1363]
Hi,

While upgrading to spring boot `2.5.5`, I'm facing the following exception:
```
java.util.NoSuchElementException: null
	at java.base/java.util.Arrays$ArrayItr.next(Arrays.java:4431)
	at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.lambda$linkTo$1(WebMvcLinkBuilderFactory.java:128)
	at org.springframework.hateoas.server.core.WebHandler.lambda$linkTo$0(WebHandler.java:156)
	at org.springframework.hateoas.server.core.WebHandler.linkTo(WebHandler.java:86)
	at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.linkTo(WebMvcLinkBuilderFactory.java:121)
	at org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo(WebMvcLinkBuilder.java:165)
	at org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo(WebMvcLinkBuilder.java:139)
	at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.linkTo(WebMvcLinkBuilderFactory.java:108)
	at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.linkTo(WebMvcLinkBuilderFactory.java:58)
	at com.blazebit.persistence.spring.hateoas.webmvc.KeysetAwarePagedResourcesAssemblerArgumentResolver.resolveBaseUri(KeysetAwarePagedResourcesAssemblerArgumentResolver.java:86)
	at com.blazebit.persistence.spring.hateoas.webmvc.KeysetAwarePagedResourcesAssemblerArgumentResolver.resolveArgument(KeysetAwarePagedResourcesAssemblerArgumentResolver.java:68)
	at org.springframework.web.method.support.HandlerMethodArgumentResolverComposite.resolveArgument(HandlerMethodArgumentResolverComposite.java:121)
```
This is probably related to a change in spring-hateoas `1.3.4`, I _think_ introduced by https://github.com/spring-projects/spring-hateoas/issues/1589
Could you please have a look?
Thanks!

**Comment by beikov** [comment]
Hi @ask4gilles and thanks for the report. I'll look into it as soon as possible, but as always, a pre-analysis or suggestions how to fix this would greatly help!

**Comment by ask4gilles** [comment]
That was fast! I'm debugging ;-)

**Comment by ask4gilles** [comment]
@beikov I'm able to reproduce it by changing the `spring.hateoas` version in the `spring-data-2.5.x`profile

```
<profile>
<id>spring-data-2.5.x</id>
<properties>
<version.spring.hateoas>1.3.4</version.spring.hateoas>
</properties>
```
`$ mvn test -P spring-data-2.5.x
`
```
[INFO] Running com.blazebit.persistence.spring.hateoas.webmvc.DocumentControllerTest
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Sep 23, 2021 2:07:31 PM com.blazebit.persistence.impl.CriteriaBuilderConfigurationProviderImpl createConfiguration
INFO: Blaze-Persistence version: Blaze-Persistence 'Lockjaw' 1.6.2-SNAPSHOT
Sep 23, 2021 2:07:31 PM com.blazebit.annotation.AnnotationUtils <clinit>
WARNING: Stereotype annotation can not be found, skipping annotation inheritance via stereotype.
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 6.358 s <<< FAILURE! - in com.blazebit.persistence.spring.hateoas.webmvc.DocumentControllerTest
[ERROR] com.blazebit.persistence.spring.hateoas.webmvc.DocumentControllerTest.testDocumentControllerHateoas  Time elapsed: 5.879 s  <<< ERROR!
org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.util.NoSuchElementException
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:645)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
        at org.springframework.test.web.servlet.TestDispatcherServlet.service(TestDispatcherServlet.java:72)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
        at org.springframework.mock.web.MockFilterChain$ServletFilterProxy.doFilter(MockFilterChain.java:167)
        at org.springframework.mock.web.MockFilterChain.doFilter(MockFilterChain.java:134)
        at org.springframework.test.web.servlet.MockMvc.perform(MockMvc.java:183)
        at com.blazebit.persistence.spring.hateoas.webmvc.DocumentControllerTest.testDocumentControllerHateoas(DocumentControllerTest.java:71)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
        at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:157)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
        at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: java.util.NoSuchElementException
        at java.base/java.util.Arrays$ArrayItr.next(Arrays.java:4431)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.lambda$linkTo$1(WebMvcLinkBuilderFactory.java:128)
        at org.springframework.hateoas.server.core.WebHandler.lambda$linkTo$0(WebHandler.java:156)
        at org.springframework.hateoas.server.core.WebHandler.linkTo(WebHandler.java:86)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.linkTo(WebMvcLinkBuilderFactory.java:121)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo(WebMvcLinkBuilder.java:165)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo(WebMvcLinkBuilder.java:139)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.linkTo(WebMvcLinkBuilderFactory.java:108)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.linkTo(WebMvcLinkBuilderFactory.java:58)
        at com.blazebit.persistence.spring.hateoas.webmvc.KeysetAwarePagedResourcesAssemblerArgumentResolver.resolveBaseUri(KeysetAwarePagedResourcesAssemblerArgumentResolver.java:86)
        at com.blazebit.persistence.spring.hateoas.webmvc.KeysetAwarePagedResourcesAssemblerArgumentResolver.resolveArgument(KeysetAwarePagedResourcesAssemblerArgumentResolver.java:68)
        at org.springframework.web.method.support.HandlerMethodArgumentResolverComposite.resolveArgument(HandlerMethodArgumentResolverComposite.java:121)
        at org.springframework.web.method.support.InvocableHandlerMethod.getMethodArgumentValues(InvocableHandlerMethod.java:170)
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:137)
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:106)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1064)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
        ... 48 more

[ERROR] com.blazebit.persistence.spring.hateoas.webmvc.DocumentControllerTest.testDocumentControllerHeaders  Time elapsed: 0.458 s  <<< ERROR!
org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.util.NoSuchElementException
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:645)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
        at org.springframework.test.web.servlet.TestDispatcherServlet.service(TestDispatcherServlet.java:72)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
        at org.springframework.mock.web.MockFilterChain$ServletFilterProxy.doFilter(MockFilterChain.java:167)
        at org.springframework.mock.web.MockFilterChain.doFilter(MockFilterChain.java:134)
        at org.springframework.test.web.servlet.MockMvc.perform(MockMvc.java:183)
        at com.blazebit.persistence.spring.hateoas.webmvc.DocumentControllerTest.testDocumentControllerHeaders(DocumentControllerTest.java:59)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
        at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:157)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
        at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: java.util.NoSuchElementException
        at java.base/java.util.Arrays$ArrayItr.next(Arrays.java:4431)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.lambda$linkTo$1(WebMvcLinkBuilderFactory.java:128)
        at org.springframework.hateoas.server.core.WebHandler.lambda$linkTo$0(WebHandler.java:156)
        at org.springframework.hateoas.server.core.WebHandler.linkTo(WebHandler.java:86)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.linkTo(WebMvcLinkBuilderFactory.java:121)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo(WebMvcLinkBuilder.java:165)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo(WebMvcLinkBuilder.java:139)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.linkTo(WebMvcLinkBuilderFactory.java:108)
        at org.springframework.hateoas.server.mvc.WebMvcLinkBuilderFactory.linkTo(WebMvcLinkBuilderFactory.java:58)
        at com.blazebit.persistence.spring.hateoas.webmvc.KeysetAwarePagedResourcesAssemblerArgumentResolver.resolveBaseUri(KeysetAwarePagedResourcesAssemblerArgumentResolver.java:86)
        at com.blazebit.persistence.spring.hateoas.webmvc.KeysetAwarePagedResourcesAssemblerArgumentResolver.resolveArgument(KeysetAwarePagedResourcesAssemblerArgumentResolver.java:68)
        at org.springframework.web.method.support.HandlerMethodArgumentResolverComposite.resolveArgument(HandlerMethodArgumentResolverComposite.java:121)
        at org.springframework.web.method.support.InvocableHandlerMethod.getMethodArgumentValues(InvocableHandlerMethod.java:170)
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:137)
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:106)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1064)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
        ... 48 more

[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   DocumentControllerTest.testDocumentControllerHateoas:71 ╗ NestedServlet Reques...
[ERROR]   DocumentControllerTest.testDocumentControllerHeaders:59 ╗ NestedServlet Reques...

```

Here is the culprit :) : https://github.com/spring-projects/spring-hateoas/blob/1.3.x/src/main/java/org/springframework/hateoas/server/mvc/WebMvcLinkBuilderFactory.java#L128

Hope it helps.

**Comment by beikov** [comment]
Thanks for the analysis. Do you think this might be a bug of Spring HATEOAS? AFAICT this is done to determine the baseUri for the request, but maybe there is an easier way to do this?

**Comment by ask4gilles** [comment]
From what I see there, the assumption is made that the `parameterValues` are never empty, but in our case, it is...
I don't know if @odrotbohm could chime in? :)

**Comment by odrotbohm** [comment]
Happy to help! At first glance, it's weird that you apparently don't hand in parameter values for a target controller method that actually takes parameters. We could of course default to assume null for the ones missing but it feels like this would mask issues better solved on the call side?

**Comment by beikov** [comment]
Hi @odrotbohm,

it's not that I explicitly omit or pass null values, but I call `MethodLinkBuilderFactory#linkTo(Class<?>, Method)` as you can see in https://github.com/Blazebit/blaze-persistence/blob/master/integration/spring-hateoas/webmvc/src/main/java/com/blazebit/persistence/spring/hateoas/webmvc/KeysetAwarePagedResourcesAssemblerArgumentResolver.java#L86

It seems that Spring HATEOAS is then trying to figure out the values somehow but receives empty `parameterValues`.

**Comment by odrotbohm** [comment]
Ouch, that's most certainly a bug. Should be easy to reproduce. I'll keep you posted.

**Comment by odrotbohm** [comment]
An easy workaround should be to rather call `MethodLinkBuilderFactory.linkTo(type, method, new Object[method.getParameterTypes().length]`.

**Comment by odrotbohm** [comment]
I've filed spring-projects/spring-hateoas#1652 and have a fix for that that solves the problem properly in 1.4. For the backport to 1.3.x I am inclined to just reject a mismatch in parameters and parameter values' length, so that we at least fail with a better error message that indicates where the problem is. In 1.4 a new overload would make your original code work again, as that allows us to explicitly distinguish between cases in which parameters were given explicitly or explicitly not.

Does that help?

**Comment by beikov** [comment]
Hey @odrotbohm,

thanks for the quick action and nice to hear that this is fixed in 1.4. I'll add this "workaround" :)
